File: nexus-adapter-sdks/nexus-adapter-sdk-ts/src/monitor.ts
import { AdapterLogger } from "./logger.js";
import { NexusEvent } from "./protocol.js";
import type { AdapterContext, AdapterDefinition } from "./run.js";

export type EmitFunc = (event: NexusEvent) => void;

export type PollConfig = {
  intervalMs: number;
  fetch: (
    ctx: AdapterContext,
    since: Date,
    account: string,
  ) => Promise<{ events: NexusEvent[]; newCursor?: Date }> | { events: NexusEvent[]; newCursor?: Date };
  initialCursor?: Date;
  errorBackoffMs?: number;
  maxConsecutiveErrors?: number;
};

export type MonitorHandler = NonNullable<AdapterDefinition["monitor"]>;

export function pollMonitor(config: PollConfig): MonitorHandler {
  return async (ctx: AdapterContext, args: { account: string }, emit: EmitFunc) => {
    const intervalMs = config.intervalMs;
    if (!Number.isFinite(intervalMs) || intervalMs <= 0) {
      throw new Error(`pollMonitor: invalid intervalMs: ${intervalMs}`);
    }

    let cursor = config.initialCursor ?? new Date();
    let errorBackoffMs = config.errorBackoffMs ?? intervalMs;
    if (!Number.isFinite(errorBackoffMs) || errorBackoffMs <= 0) {
      errorBackoffMs = intervalMs;
    }

    let consecutiveErrors = 0;
    const maxConsecutiveErrors = config.maxConsecutiveErrors ?? 0;

    while (true) {
      if (ctx.signal.aborted) {
        ctx.log.info("monitor shutting down (signal aborted)");
        return;
      }

      let fetched: { events: NexusEvent[]; newCursor?: Date };
      try {
        fetched = await config.fetch(ctx, cursor, args.account);
      } catch (err) {
        consecutiveErrors++;
        ctx.log.error("poll fetch error (%d consecutive): %s", consecutiveErrors, errorToString(err));

        if (maxConsecutiveErrors > 0 && consecutiveErrors >= maxConsecutiveErrors) {
          throw err instanceof Error ? err : new Error(String(err));
        }

        await sleep(errorBackoffMs, ctx.signal, ctx.log);
        continue;
      }

      consecutiveErrors = 0;

      const events = fetched.events ?? [];
      for (const event of events) {
        emit(event);
      }
      if (events.length > 0) {
        ctx.log.debug("emitted %d events", events.length);
      }

      if (fetched.newCursor instanceof Date && !Number.isNaN(fetched.newCursor.valueOf())) {
        cursor = fetched.newCursor;
      }

      await sleep(intervalMs, ctx.signal, ctx.log);
    }
  };
}

function errorToString(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}

async function sleep(ms: number, signal: AbortSignal, log: AdapterLogger): Promise<void> {
  await new Promise<void>((resolve) => {
    const t = setTimeout(() => {
      signal.removeEventListener("abort", onAbort);
      resolve();
    }, ms);

    const onAbort = () => {
      clearTimeout(t);
      signal.removeEventListener("abort", onAbort);
      log.debug("sleep aborted");
      resolve();
    };

    if (signal.aborted) {
      clearTimeout(t);
      resolve();
      return;
    }
    signal.addEventListener("abort", onAbort, { once: true });
  });
}

File: nexus-adapter-sdks/nexus-adapter-sdk-go/monitor.go
package nexadapter

import (
	"context"
	"fmt"
	"time"
)

// EmitFunc is called by monitor/backfill handlers to emit NexusEvents.
// The SDK handles JSONL serialization and writing to stdout. Thread-safe.
type EmitFunc func(event NexusEvent)

// PollConfig configures a polling-based monitor.
//
// Use PollMonitor when your adapter reads from a database, API, or any
// other pull-based source. The SDK manages the polling loop, cursor
// advancement, and JSONL emission — you just write the fetch function.
type PollConfig struct {
	// Interval between polls. Required.
	Interval time.Duration

	// Fetch retrieves events since the given cursor time.
	// Returns the events and the new cursor position.
	// If newCursor is zero, the cursor is not advanced (useful when no events found).
	// The function should respect ctx cancellation.
	Fetch func(ctx context.Context, since time.Time) (events []NexusEvent, newCursor time.Time, err error)

	// InitialCursor is the starting cursor position.
	// Defaults to time.Now() if zero.
	InitialCursor time.Time

	// ErrorBackoff is how long to wait after a fetch error before retrying.
	// Defaults to the poll Interval if zero.
	ErrorBackoff time.Duration

	// MaxConsecutiveErrors is how many consecutive fetch errors before the
	// monitor exits with an error. 0 means never exit (keep retrying forever).
	MaxConsecutiveErrors int
}

// PollMonitor returns a MonitorFunc that polls at a fixed interval.
// It handles the sleep/poll/emit/cursor-advance cycle. The adapter author
// only provides the Fetch function.
//
// Example:
//
//	adapter := nexadapter.Adapter{
//	    Monitor: nexadapter.PollMonitor(nexadapter.PollConfig{
//	        Interval: 10 * time.Second,
//	        Fetch: func(ctx context.Context, since time.Time) ([]NexusEvent, time.Time, error) {
//	            rows := queryMessagesAfter(since)
//	            var events []NexusEvent
//	            var latest time.Time
//	            for _, row := range rows {
//	                events = append(events, convertToEvent(row))
//	                if row.Date.After(latest) {
//	                    latest = row.Date
//	                }
//	            }
//	            return events, latest, nil
//	        },
//	    }),
//	}
func PollMonitor(config PollConfig) func(ctx context.Context, account string, emit EmitFunc) error {
	return func(ctx context.Context, account string, emit EmitFunc) error {
		cursor := config.InitialCursor
		if cursor.IsZero() {
			cursor = time.Now()
		}

		errorBackoff := config.ErrorBackoff
		if errorBackoff == 0 {
			errorBackoff = config.Interval
		}

		consecutiveErrors := 0

		for {
			select {
			case <-ctx.Done():
				LogInfo("monitor shutting down (context cancelled)")
				return nil
			default:
			}

			events, newCursor, err := config.Fetch(ctx, cursor)
			if err != nil {
				consecutiveErrors++
				LogError("poll fetch error (%d consecutive): %v", consecutiveErrors, err)

				if config.MaxConsecutiveErrors > 0 && consecutiveErrors >= config.MaxConsecutiveErrors {
					return fmt.Errorf("too many consecutive errors (%d): %w", consecutiveErrors, err)
				}

				// Back off on error
				select {
				case <-ctx.Done():
					return nil
				case <-time.After(errorBackoff):
					continue
				}
			}

			// Reset error counter on success
			consecutiveErrors = 0

			// Emit events
			for _, event := range events {
				emit(event)
			}

			if len(events) > 0 {
				LogDebug("emitted %d events", len(events))
			}

			// Advance cursor
			if !newCursor.IsZero() {
				cursor = newCursor
			}

			// Wait for next poll
			select {
			case <-ctx.Done():
				LogInfo("monitor shutting down (context cancelled)")
				return nil
			case <-time.After(config.Interval):
			}
		}
	}
}
